server/src/main/java/org/opensearch/indices/IndicesRequestCache.java

```diff
@@ -371,40 +412,300 @@ public int hashCode() {
                     iterator.remove();
-                } else {
-                    // If the flow comes here, then we should have a open shard available on node.
-                    if (currentKeysToClean.contains(
-                        new CleanupKey(cacheEntityLookup.apply(key.shardId).orElse(null), key.readerCacheKeyId)
-                    )) {
-                        iterator.remove();
-                    }
                 }
             }
+            cache.refresh();
+        }
+
+        /**
+         * Determines whether a key should be removed from the cache.
+         *
+         * <p>This method checks if the key's shardId is present in the cleanupKeysFromClosedShards set,
+         * indicating that the shard has been closed and the key should be removed. If the shardId is not present,
+         * it checks if the key's readerCacheKeyId is present in the cleanupKeysFromOutdatedReaders set,
+         * indicating that the reader has been invalidated and the key should be removed.
+         *
+         * @param key The key to check for removal.
+         * @param cleanupKeysFromOutdatedReaders A set of CleanupKeys with open shard but invalidated readerCacheKeyId.
+         * @param cleanupKeysFromClosedShards A set of CleanupKeys of a closed shard.
+         * @return true if the key should be removed, false otherwise.
+         */
+        private synchronized boolean shouldRemoveKey(
+            Key key,
+            Set<CleanupKey> cleanupKeysFromOutdatedReaders,
+            Set<Object> cleanupKeysFromClosedShards
+        ) {
+            if (cleanupKeysFromClosedShards.contains(key.shardId)) {
+                return true;
+            } else {
+                CleanupKey cleanupKey = new CleanupKey(cacheEntityLookup.apply(key.shardId).orElse(null), key.readerCacheKeyId);
+                return cleanupKeysFromOutdatedReaders.contains(cleanupKey);
+            }
+        }
+
+        /**
+         * Determines whether the cache cleanup process can be skipped based on the staleness threshold.
+         *
+         * <p>If the percentage of stale keys is less than the provided staleness threshold returns true,
+         * indicating that the cache cleanup process can be skipped.
+         *
+         * @param cleanThresholdPercent The staleness threshold as a percentage.
+         * @return true if the cache cleanup process can be skipped, false otherwise.
+         */
+        private synchronized boolean canSkipCacheCleanup(double cleanThresholdPercent) {
+            if (cleanThresholdPercent == 0.0) {
+                return false;
+            }
+            if (staleKeysInCachePercentage() < cleanThresholdPercent) {
+                if (logger.isDebugEnabled()) {
+                    logger.debug(
+                        "Skipping cache cleanup since the percentage of stale keys is less than the threshold : " + stalenessThreshold
```

Comment:
Skipping cache clean for IndicesRequestCache .... ?
Also print both staleKeysInCachePercentage() and cleanThresholdPercent?

Reply:
agree